Add support for symbolic parameters

pepijndevos opened the pull request:
With this change you can express parameters as an expression of another parameter
```
par = [
    σ => 1,
    ρ => 0.1+σ,
    β => ρ*1.1
]
```
In combination with `flatten` this allows to express parameter dependencies across different levels, where you can define a slew of lower level parameter in terms of a single top-level definition for example.

To make this usage more convenient I have added assignment to parameters to set the default. Currently there does not appear to be an API to generate symbol names in the flattened system other than by specifying the full path with `\_+` or using the hierarchical model to get it, which might be fine.

I'm currently not checking parameters for cycles.

I'm not 100% sure if my usage of `namespace_expr` is correct because of the unused `iv` parameter.

I could not find a way to properly test the `NonlinearProblem` parameters and solution. So I've resorted to manually calling `varmap_to_vars` and just running the solver to see that nothing breaks.

YingboMa commented on src/systems/abstractsystem.jl:
```diff
@@ -181,6 +181,15 @@ function Base.getproperty(sys::AbstractSystem, name::Symbol)
     throw(error("Variable $name does not exist"))
 end
 
+function Base.setproperty!(sys::AbstractSystem, prop::Symbol, val)
```
Don't make it `Symbol` and assume it must be a `Parameter`. We can use another function like `setdefault(sys, sym, val)` to do this.

pepijndevos replied:
Isn't that what `getparameter` is already doing? If I write `sys.par` that will convert the `par` symbol and return the parameter, if it is present.

This function is so that you can write `flat.b = ns.σ*0.5` as I do in the test case. Else you have to write `flat.default_p[flat.b] = ns.σ*0.5` and IIRC it acts weird because of `Num` wrapping.

YingboMa replied:
That's not what `getproperty` is doing. It doesn't assume the name refers to a parameter. Could you add a check for states and set the `u0`, too?

YingboMa commented on src/systems/abstractsystem.jl:
```diff
@@ -181,6 +181,15 @@ function Base.getproperty(sys::AbstractSystem, name::Symbol)
+function Base.setproperty!(sys::AbstractSystem, prop::Symbol, val)
+    param = Sym{Parameter{Real}}(prop)
+    if param in parameters(sys)
+        sys.default_p[param] = val
```
This should be
```julia
sys.default_p[value(param)] = value(val)
```
to unwrap potential `Num`.

pepijndevos replied:
Could you explain this? `param` was just constructed a few lines earlier. For `val` it might make sense, I just don't fully understand `Num` wrapping. The reason for it, and when something needs to be a Num or not.

YingboMa commented on src/systems/abstractsystem.jl:
```diff
@@ -203,7 +212,7 @@ end
 function namespace_default_p(sys)
     d_p = default_p(sys)
-    Dict(parameters(sys, k) => d_p[k] for k in keys(d_p))
+    Dict(parameters(sys, k) => namespace_expr(d_p[k], nameof(sys), []) for k in keys(d_p))
```
I am not sure what this is trying to do. Let's take a step back and don't handle multiple level yet.

pepijndevos replied:
It works perfectly, the only issue is that `namespace_expr` has a special case where it does not namespace the independent variable. There should be no independent variable in the parameters, so I pass a dummy value.

Imagine you have a parameter dict like this inside a nested system
```
par = [
    σ => 1,
    ρ => 0.1+σ,
    β => ρ*1.1
]
```
If you call `default_p` on this, it will return
```
par = [
    sys₊σ => 1,
    sys₊ρ => 0.1+σ,
    sys₊β => ρ*1.1
]
```
Which is wrong because now  `σ` doesn't resolve anymore. The correct dict is
```
par = [
    sys₊σ => 1,
    sys₊ρ => 0.1+sys₊σ,
    sys₊β => sys₊ρ*1.1
]
```
Any this is what `namespace_expr` accomplishes. You always have to namespace the keys together with the values to maintain correctness.

pepijndevos commented:
I ran into a problem with this PR, which is that if you don't pass any explicit parameters, it uses a `NullParameters`, and does not call `varmap_to_vars` at all. This then leaks symbolic expressions from the defaults into the solver.

I worked around this in my code by passing `Pair[]` as the parameters which makes it call `varmap_to_vars`, but the problem is that if there are no parameters at all, it will fail at `Base.promote_typeof` on `Any`.

YingboMa commented on test/symbolic_parameters.jl:
```diff
@@ -0,0 +1,35 @@
+flatres = ModelingToolkit.varmap_to_vars(Dict(), parameters(flat), defaults=ModelingToolkit.default_p(flat))
+@test flatres == [0.5, 1, 0.1+1, (0.1+1)*1.1]
+
+prob = NonlinearProblem(flat, ones(4), Pair[])
```
Could you add another test for the
```julia
prob = NonlinearProblem(flat, ones(4))
```
case?

pepijndevos commented:
Thanks for the review.

I've updated the PR to properly handle u0 defaults as well, and added this to the testcase.
And also added a basic test of the NullParameters.

YingboMa reviewed (approved):
LGTM!

ChrisRackauckas commented on test/symbolic_parameters.jl:
```diff
@@ -0,0 +1,43 @@
+top = NonlinearSystem([0 ~ -a + ns.x+b], [a], [b], systems=[ns], name=:top)
+top.b = ns.σ*0.5
+
+res = ModelingToolkit.varmap_to_vars(Dict(), parameters(top), defaults=ModelingToolkit.default_p(top))
```
can we get a test on the defaults with the component notation?

pepijndevos replied:
What do you mean?

I was a bit puzzled by testing parameters with `NonlinearProblem` because you can not easily access the final parameter list.
I guess I could add a test on the computer problem u0, which you can access it appears.
Is that what you mean?

ChrisRackauckas replied:
I mean the high level overriding the default with like `ns.σ => b`

pepijndevos replied:
Good point! It's actually running into a problem where `in` compares the states and... returns a symbolic expression rather than a boolean.
```
julia> s = Sym{ModelingToolkit.Parameter{Real}}(:foo)
foo

julia> s == s
true

julia> s = Sym{Real}(:foo)
foo

julia> s == s
foo == foo
```

pepijndevos replied:
Is there like a `in` that uses `===`? That one seems to work (was just a wild guess based on JS experience)
Other alternative is to compare `s.name` with a custom HOF.

pepijndevos commented:
Okay I got the states figured out, and added more tests.

One thing I ran into a few times is when Julia infers the type of a datastructure, and later complains when you try to add a different type. I managed to solve this a few times by adding the right type annotations, and other cases I worked around it by purposefully making it infer to Any.

pbouffard commented:
Friendly appeal to some updates to docs explaining this cool new feature 😄 .
